grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ResolutionPlan.java

```diff
@@ -0,0 +1,214 @@
+                .filter(Objects::nonNull)
+                .filter(properties::contains)
+                .distinct()
+                .map(propertyMap::get)
+                .collect(Collectors.toCollection(LinkedList::new));
+
+        Set<Atom> nonResolvableAtoms = new HashSet<>();
+        while (!atoms.isEmpty()) {
+            Atom top = atoms.getFirst();
+            atoms.remove(top);
+            if (top.isRuleResolvable()) {
+                if (!nonResolvableAtoms.isEmpty()) {
+                    queries.add(ReasonerQueries.create(nonResolvableAtoms, graph));
+                    nonResolvableAtoms.clear();
+                }
+                queries.add(new ReasonerAtomicQuery(top));
+            } else {
+                nonResolvableAtoms.add(top);
+                if (atoms.isEmpty()) queries.add(ReasonerQueries.create(nonResolvableAtoms, graph));
+            }
+        }
+        return queries;
+    }
+
+    /**
+     * compute the resolution plan - list of atomic queries ordered by their resolution priority
+     * @return list of prioritised queries
+     */
+    static LinkedList<ReasonerQueryImpl> getResolutionPlan(ReasonerQueryImpl query){
+        LinkedList<ReasonerQueryImpl> queries = new LinkedList<>();
+        GraknGraph graph = query.graph();
+
+        LinkedList<Atom> atoms = query.selectAtoms().stream()
+                .sorted(Comparator.comparing(at -> -at.baseResolutionPriority()))
+                .collect(Collectors.toCollection(LinkedList::new));
+
+        Atom top = atoms.getFirst();
+        Set<Atom> nonResolvableAtoms = new HashSet<>();
+        Set<Var> subbedVars = query.getIdPredicates().stream().map(IdPredicate::getVarName).collect(Collectors.toSet());
+        while (!atoms.isEmpty()) {
+
+            subbedVars.addAll(top.getVarNames());
+            atoms.remove(top);
+
+            if (top.isRuleResolvable()) {
+                if (!nonResolvableAtoms.isEmpty()){
+                    queries.add(ReasonerQueries.create(nonResolvableAtoms, graph));
+                    nonResolvableAtoms.clear();
+                }
+                queries.add(new ReasonerAtomicQuery(top));
+            } else {
+                nonResolvableAtoms.add(top);
+                if (atoms.isEmpty()) queries.add(ReasonerQueries.create(nonResolvableAtoms, graph));
+            }
+
+            //look at neighbours up to two hops away
+            top = top.getNeighbours().filter(atoms::contains)
+                    .flatMap(at -> Stream.concat(Stream.of(at), at.getNeighbours().filter(atoms::contains)))
+                    .sorted(Comparator.comparing(at -> -at.computePriority(subbedVars)))
+                    .findFirst().orElse(null);
+
+            //top is disconnected atom
+            if (top == null) {
+                top = atoms.stream()
+                        .sorted(Comparator.comparing(at -> -at.computePriority(subbedVars)))
+                        .findFirst().orElse(null);
```

Comment:
I'd call `get()` instead of `orElse(null)` in this case. If there's nothing present, we actually want to crash immediately (since we try and access `top` in the next iteration).

IntelliJ gives a warning about using `get()` but that's just because it has a bad name :P